File: MPI_on_stampede/linresid.f90
! -*- Mode: Fortran90; -*-
!-----------------------------------------------------------------
! Daniel R. Reynolds
! SMU, Mathematics
! Math 6370
! 23 March 2009
!=================================================================


subroutine linresid(a, b, c, u, r, res, norm2, locN, N, comm, ierr)
  !-----------------------------------------------------------------
  ! Description: 
  !    calculates the linear residual and its averaged 2-norm (WRMS)
  !-----------------------------------------------------------------
  !======= Inclusions ===========
  use mpi

  !======= Declarations =========
  implicit none

  integer, intent(in) :: locN, N, comm
  integer, intent(out) :: ierr
  double precision, intent(in),  dimension(locN) :: a, b, c, u, r
  double precision, intent(out), dimension(locN) :: res
  double precision, intent(out) :: norm2
  double precision :: u_l, u_r, s_l, s_r
  double precision :: tmp
  integer :: my_id, nprocs, k
  integer :: status(MPI_STATUS_SIZE)
  integer :: id_recv_l, id_recv_r, id_send_l, id_send_r

  !======= Internals ============
  
  ! Get MPI parallelism information from comm
  call MPI_Comm_size(comm, nprocs, ierr)
  call MPI_Comm_rank(comm, my_id, ierr)

  ! open asynchronous receive channels for neighbor boundary values
  u_l = 0.d0
  u_r = 0.d0
  if (my_id /= 0) then
     call MPI_Irecv(u_l, 1, MPI_DOUBLE_PRECISION, my_id-1, 100, &
          comm, id_recv_l, ierr)
     if (ierr /= 0) then
        print *,'linresid error: MPI_Irecv failed with error = ',ierr
        return
     end if
  end if
  if (my_id /= nprocs-1) then
     call MPI_Irecv(u_r, 1, MPI_DOUBLE_PRECISION, my_id+1, 101, &
          comm, id_recv_r, ierr)
     if (ierr /= 0) then
        print *,'linresid error: MPI_Irecv failed with error = ',ierr
        return
     end if
  end if

  ! send boundary values to neighbor processes
  s_l = u(1)
  s_r = u(locN)
  if (my_id /= nprocs-1) then
     call MPI_Isend(s_r, 1, MPI_DOUBLE_PRECISION, my_id+1, 100, &
          comm, id_send_r, ierr)
     if (ierr /= 0) then
        print *,'linresid error: MPI_Isend failed with error = ',ierr
        return
     end if
  end if
  if (my_id /= 0) then
     call MPI_Isend(s_l, 1, MPI_DOUBLE_PRECISION, my_id-1, 101, &
          comm, id_send_l, ierr)
     if (ierr /= 0) then
        print *,'linresid error: MPI_Isend failed with error = ',ierr
        return
     end if
  end if

  ! wait for left boundary value to arrive before using
  if (my_id /= 0) then
     call MPI_Wait(id_recv_l, status, ierr)
     if (ierr /= 0) then
        print *,'linresid error: MPI_Wait failed with error = ',ierr
        return
     end if
  end if

  ! compute linear residual at left of local subdomain
  res(1) = a(1)*u_l + b(1)*u(1) + c(1)*u(2) - r(1)
  norm2 = res(1)**2

  ! compute linear residual in interior of subdomain
  do k=2,locN-1
    res(k) = a(k)*u(k-1) + b(k)*u(k) + c(k)*u(k+1) - r(k)
    norm2 = norm2 + res(k)**2
  end do

  ! wait for right boundary value to arrive before using
  if (my_id /= nprocs-1) then
     call MPI_Wait(id_recv_r, status, ierr)
     if (ierr /= 0) then
        print *,'linresid error: MPI_Wait failed with error = ',ierr
        return
     end if
  end if

  ! compute linear residual at right end of local subdomain
  k = locN
  res(k) = a(k)*u(k-1) + b(k)*u(k) + c(k)*u_r - r(k)
  norm2 = norm2 + res(k)**2

  ! combine local 2-norms into global averaged 2-norm
  call MPI_Allreduce(norm2, tmp, 1, MPI_DOUBLE_PRECISION, &
       MPI_SUM, comm, ierr)
  if (ierr /= 0) then
     print *,'linresid error: MPI_Allreduce failed with error = ',ierr
     return
  end if
  norm2 = sqrt(tmp/N)

  return

end subroutine linresid
!=================================================================

File: MPI_on_stampede/linresid.cpp
/* Daniel R. Reynolds
   SMU Mathematics
   Math 6370
   28 February 2013 */


// Inclusions
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "mpi.h"


/* Description: 
      calculates the linear residual and returns its averaged 2-norm (WRMS) */
double linresid(double *a, double *b, double *c, double *u, double *r, 
		double *res, int local_N, int global_N, MPI_Comm comm) {

  // declarations
  int ierr, nprocs, my_id, k;
  double u_l, u_r, s_l, s_r, tmp, norm2;
  MPI_Request id_recv_l, id_recv_r, id_send_l, id_send_r;
  MPI_Status status;

  // Get MPI parallelism information from comm
  ierr = MPI_Comm_size(comm, &nprocs);
  if (ierr != 0) {
    fprintf(stderr,"error in MPI_Comm_size = %i\n",ierr);
    MPI_Abort(comm, 1);
  }
  ierr = MPI_Comm_rank(comm, &my_id);
  if (ierr != 0) {
    fprintf(stderr,"error in MPI_Comm_rank = %i\n",ierr);
    MPI_Abort(comm, 1);
  }

  // open asynchronous receive channels for neighbor boundary values
  u_l = 0.0;
  if (my_id > 0) {
    ierr = MPI_Irecv(&u_l, 1, MPI_DOUBLE, my_id-1, 100, comm, &id_recv_l);
    if (ierr != 0) {
      fprintf(stderr,"error in MPI_Irecv = %i\n",ierr);
      MPI_Abort(comm, 1);
    }
  } 
  u_r = 0.0;
  if (my_id < nprocs-1) {
    ierr = MPI_Irecv(&u_r, 1, MPI_DOUBLE, my_id+1, 101, comm, &id_recv_r);
    if (ierr != 0) {
      fprintf(stderr,"error in MPI_Irecv = %i\n",ierr);
      MPI_Abort(comm, 1);
    }
  }

  // send boundary values to neighbor processes
  s_r = u[local_N-1];
  if (my_id < nprocs-1) {
    ierr = MPI_Isend(&s_r, 1, MPI_DOUBLE, my_id+1, 100, comm, &id_send_r);
    if (ierr != 0) {
      fprintf(stderr,"error in MPI_Isend = %i\n",ierr);
      MPI_Abort(comm, 1);
    }
  } 
  s_l = u[0];
  if (my_id > 0) {
    ierr = MPI_Isend(&s_l, 1, MPI_DOUBLE, my_id-1, 101, comm, &id_send_l);
    if (ierr != 0) {
      fprintf(stderr,"error in MPI_Isend = %i\n",ierr);
      MPI_Abort(comm, 1);
    }
  }

  // compute linear residual in interior of subdomain
  norm2 = 0.0;
  for (k=1; k<local_N-1; k++) {
    res[k] = a[k]*u[k-1] + b[k]*u[k] + c[k]*u[k+1] - r[k];
    norm2 += res[k]*res[k];
  }

  // wait for left boundary value to arrive before using
  if (my_id > 0) {
    ierr = MPI_Wait(&id_recv_l, &status);
    if (ierr != 0) {
      fprintf(stderr,"error in MPI_Wait = %i\n",ierr);
      MPI_Abort(comm, 1);
    }
  }

  // wait for right boundary value to arrive before using
  if (my_id < nprocs-1) {
    ierr = MPI_Wait(&id_recv_r, &status);
    if (ierr != 0) {
      fprintf(stderr,"error in MPI_Wait = %i\n",ierr);
      MPI_Abort(comm, 1);
    }
  }

  // compute linear residual at left of subdomain
  res[0] = a[0]*u_l + b[0]*u[0] + c[0]*u[1] - r[0];
  norm2 += res[0]*res[0];

  // compute linear residual at right end of subdomain
  k = local_N-1;
  res[k] = a[k]*u[k-1] + b[k]*u[k]+ c[k]*u_r - r[k];
  norm2 += res[k]*res[k];

  // combine local 2-norms into global averaged 2-norm
  tmp = 0.0;
  ierr = MPI_Allreduce(&norm2, &tmp, 1, MPI_DOUBLE, MPI_SUM, comm);
  if (ierr != 0) {
    fprintf(stderr,"error in MPI_Allreduce = %i\n",ierr);
    MPI_Abort(comm, 1);
  }
  norm2 = sqrt(tmp/global_N);


  // wait for sends to finish
  if (my_id > 0) {
    ierr = MPI_Wait(&id_send_l, &status);
    if (ierr != 0) {
      fprintf(stderr,"error in MPI_Wait = %i\n",ierr);
      MPI_Abort(comm, 1);
    }
  }

  // wait for right boundary value to arrive before using
  if (my_id < nprocs-1) {
    ierr = MPI_Wait(&id_send_r, &status);
    if (ierr != 0) {
      fprintf(stderr,"error in MPI_Wait = %i\n",ierr);
      MPI_Abort(comm, 1);
    }
  }

  return norm2;

} // end linresid
